lib/tpm/validate.go

```diff
@@ -0,0 +1,201 @@
+	// CAs.
+	EKCertVerified bool `json:"ek_cert_verified"`
+}
+
+// JoinAuditAttributes returns a series of attributes that can be inserted into
+// audit events related to a specific join.
+func (c *ValidatedTPM) JoinAuditAttributes() (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"ek_pub_hash":      c.EKPubHash,
+		"ek_cert_serial":   c.EKCertSerial,
+		"ek_cert_verified": c.EKCertVerified,
+	}, nil
+}
+
+// Validate takes the parameters from a remote TPM and performs the necessary
+// initial checks before then generating an encrypted credential challenge for
+// the client to solve in a credential activation ceremony. This allows us to
+// verify that the client possesses the TPM corresponding to the EK public key
+// or certificate presented by the client.
+func Validate(
+	ctx context.Context, log *slog.Logger, params ValidateParams,
+) (*ValidatedTPM, error) {
+	ctx, span := tracer.Start(ctx, "Validate")
+	defer span.End()
+
+	// Validate params
+	switch {
+	case params.Solve == nil:
+		return nil, trace.BadParameter("solve must be non-nil")
+	case params.EKCert == nil && params.EKKey == nil:
+		return nil, trace.BadParameter("at least one of EKCert or EKKey must be provided")
+	}
+
+	ekCert, ekPub, err := parseEK(ctx, params)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	validated := &ValidatedTPM{}
+	ekPubPKIX, err := x509.MarshalPKIXPublicKey(ekPub)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	validated.EKPubHash, err = hashEKPub(ekPubPKIX)
+	if err != nil {
+		return validated, trace.Wrap(err, "hashing EK public key")
+	}
+	if ekCert != nil {
+		validated.EKCertSerial = serialString(ekCert.SerialNumber)
+	}
+
+	if params.AllowedCAs != nil {
+		if err := verifyEKCert(ctx, params.AllowedCAs, ekCert); err != nil {
+			return validated, trace.Wrap(err, "verifying EK cert")
+		}
+		validated.EKCertVerified = true
+	}
+
+	activationParameters := attest.ActivationParameters{
+		TPMVersion: attest.TPMVersion20,
+		AK:         params.AttestParams,
+		EK:         ekPub,
+	}
+	// The generate method completes initial validation that provides the
+	// following assurances:
+	// - The attestation key is of a secure length
+	// - The attestation key is marked as created within a TPM
+	// - The attestation key is marked as restricted (e.g cannot be used to
+	//   sign or decrypt external data)
+	// When the returned challenge is solved by the TPM using ActivateCredential
+	// the following additional assurance is given:
+	// - The attestation key resides in the same TPM as the endorsement key
+	solution, encryptedCredential, err := activationParameters.Generate()
+	if err != nil {
+		return validated, trace.Wrap(err, "generating credential activation challenge")
+	}
+	clientSolution, err := params.Solve(encryptedCredential)
+	if err != nil {
+		return validated, trace.Wrap(err, "asking client to perform credential activation")
+	}
+	if subtle.ConstantTimeCompare(clientSolution, solution) == 0 {
```

Comment:
Just to be extra safe?

```suggestion
	if subtle.ConstantTimeCompare(clientSolution, solution) != 1 {
```